src/subsidy.cpp

```diff
@@ -385,21 +385,17 @@ bool FindSubsidyIndustryCargoRoute()
 	CargoID cid;
 
 	/* Randomize cargo type */
-	int num_cargos = 0;
-	uint cargo_index;
-	for (cargo_index = 0; cargo_index < lengthof(src_ind->produced_cargo); cargo_index++) {
-		if (IsValidCargoID(src_ind->produced_cargo[cargo_index])) num_cargos++;
-	}
+	int num_cargos = std::count_if(std::begin(src_ind->produced), std::end(src_ind->produced), [](const auto &p) { return IsValidCargoID(p.cargo); });
 	if (num_cargos == 0) return false; // industry produces nothing
 	int cargo_num = RandomRange(num_cargos) + 1;
-	for (cargo_index = 0; cargo_index < lengthof(src_ind->produced_cargo); cargo_index++) {
-		if (IsValidCargoID(src_ind->produced_cargo[cargo_index])) cargo_num--;
-		if (cargo_num == 0) break;
-	}
-	assert(cargo_num == 0); // indicates loop didn't break as intended
-	cid = src_ind->produced_cargo[cargo_index];
-	trans = src_ind->last_month_pct_transported[cargo_index];
-	total = src_ind->last_month_production[cargo_index];
+
+	auto it = std::begin(src_ind->produced);
+	for (/* nothing */; it != std::end(src_ind->produced) && --cargo_num > 0; ++it) {}
```

Comment:
```suggestion
	auto it = std::advance(std::begin(src_ind->produced), cargo_num);
```
However, I think this code does not behave as before. This advances the first `cargo_num` entries in the array, whereas it advanced the first `cargo_num` *valid* cargoes in the new code.